### ISSUE TITLE
Support disposable pattern for `ref` structs

### ISSUE THREAD

**Issue by auduchinok** [fsharp/fslang-suggestions#1229]
I propose we support disposable pattern for [`ref` structs](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/ref-struct):
> You can define a disposable ref struct. To do that, ensure that a ref struct fits the [disposable pattern](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#pattern-based-using). That is, it has an instance or extension Dispose method, which is accessible, parameterless and has a void return type.

In recent C# versions this pattern can be used instead of implementing `IDisposable` in a struct type. If such a change to a struct type is done in a C# project/library, then this type can't be used in `use` binding in F# anymore, i.e this code:
```fsharp
use x = Library.Method()
...
```
has to be rewritten into this:
```fsharp
let x = Library.Method()
...
x.Dispose()
```

The type change doesn't need any changes in referencing C# projects, so library authors may break usages in F# code without realising it.

## Pros and Cons 

The advantages of making this adjustment to F# are
* being compatible with a recent C# feature that may be used in referenced libraries/projects; otherwise F# usages may get broken

The disadvantages of making this adjustment to F# are
* it needs work

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S-M

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [ ] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by Xyncgas** [comment]
what if we just automatically call dispose when these value go out of scope. Unlike objects they don't live longer than where they are

**Comment by vzarytovskii** [comment]
> what if we just automatically call dispose when these value go out of scope. Unlike objects they don't live longer than where they are

It is too implicit.

**Comment by edgarfgp** [comment]
@dsyme, @vzarytovskii  Any thoughts on this suggestion ?

**Comment by Xyncgas** [comment]
>It is too implicit

ref struct is going to be expiring where they are created, it's a feature people chose to use, so maybe it wouldn't be too implicit but rather natural to see ref struct's dispose being called

**Comment by vzarytovskii** [comment]
> @dsyme, @vzarytovskii  Any thoughts on this suggestion ? 

It feels a bit weird tbh, it's a very ad-hoc solution from the C# side, ducktyped dispose, but only for ref structs. 

Personally I think that we either should fully support it on any type (like we do with fixed) or wait until ref structs support interfaces (https://github.com/dotnet/csharplang/blob/main/proposals/ref-struct-interfaces.md).

I would personally vote for the latter, since it makes more sense from the design of the IL types and how disposables work already.

Tracking item for it (and other .NET 9 features) is here: https://github.com/dotnet/fsharp/issues/16967

**Comment by auduchinok** [comment]
> It feels a bit weird tbh, it's a very ad-hoc solution from the C# side, ducktyped dispose, but only for ref structs.

I agree with this, but, unfortunately, we see people using this approach in libraries, and it become harder to consume these libraries from F# side.

**Comment by abelbraaksma** [comment]
I tend to agree with @auduchinok that we shouldn't fail when it comes to consuming other libraries. 

@vzarytovskii, it wouldn't be too hard to implement this through `use x = someRefType`, where we dispose it if it contains these methods (but not the interface), as described in the C# proposal. Doing so gives us better feature parity.

**Comment by vzarytovskii** [comment]
> I tend to agree with @auduchinok that we shouldn't fail when it comes to consuming other libraries. 
> 
> 
> 
> @vzarytovskii, it wouldn't be too hard to implement this through `use x = someRefType`, where we dispose it if it contains these methods (but not the interface), as described in the C# proposal. Doing so gives us better feature parity.

My point was that we should either support all or only interface based disposal. Supporting it just for refs feels weird and very ad-hoc-y

**Comment by abelbraaksma** [comment]
The [referenced document](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#pattern-based-using) isn't very clear on whether having the `Dispose()` method present on any type is enough, or whether it must be a `ref` type. 

I agree, though, that it makes sense to allow this pattern for any type that exposes a public instance member named `Dispose`, of type `unit -> unit`.

**Comment by brianrourkeboll** [comment]
> The [referenced document](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/using#pattern-based-using) isn't very clear on whether having the `Dispose()` method present on any type is enough, or whether it must be a `ref` type.

C# does indeed only seem to do it for ref structs or types implementing `IDisposable`:

- [Ref struct](https://sharplab.io/#v2:D4AQTAjAsAUCDMACATgUwIYBMD2A7ANgJ4qoBmiAzgC7ICuAxlYgMIAUAlrkwG4CUsAb1iIRiBIk5MAaogC8APkTcA3MNHiQAFkQARdhQAO2Cqla85ikBACcrAEQAZbPgB0d3qpgBfWLA1hEAEVBNRFxSUQAWVZmRHpzIRhRZLEABiV0ZDiAfTk4zxT1AHYclykC0R9vIA==).
- [Regular struct](https://sharplab.io/#v2:D4AQTAjAsAUCDMACATgUwIYBMD2A7ANgJ6IDOALsgK4DGZiAwgBQCWudAbgJSwDesiAxAkSs6ANUQBeAHyJ2Abn6DhIACyIAIsxIAHbCVSNOU2SAgBORgCIAMtnwA6K50UwAvrFgqwiAIq8lAWFRRABZRnpEamM+GEF4oQAGOXRkKIB9KSjXBOUAdgyHMRzBD3cgA===).
- [Regular struct implementing `IDisposable`](https://sharplab.io/#v2:D4AQTAjAsAUCDMACATgUwIYBMD2A7ANgJ6IDOALsgK4DGZiAwgBQCWudAbgJSIBciAkgBFmJAA7YS6AEb5UsAN6xEyxAkSs6ANUQBeAHyJ2AbiUq1IACyJhYiakbd9qiAE5GAIgAy2fADp3nCYwAL6wsOZgiACKCqbKahqIALKM9IjU3IowKjmqAAyG6MjpAPq66UG5ZgDspb6alSqhIUA==).

**Comment by Xyncgas** [comment]
We can always browse compiler source code to be completely sure

**Comment by abelbraaksma** [comment]
We can also just go one step further and essentially do it the way it would work as if SRTP were conditionally applied. That is, if the `Dispose -> unit -> unit` is present as an accessible member, call it. 

It won't lead to any backward compat issues (as such types currently cannot be bound with `use`), though it may lead to people being less diligent in implementing `IDisposable` (which may have been the motivation for the narrower approach in C#).

In fact, that appears to be true. It was a conscious change during implementing to limit it to `ref struct` only: https://github.com/dotnet/roslyn/commit/3f4d93d4026ef9a9eb7b858f7afa22e480a6d3ce

**Comment by kerams** [comment]
`ref struct` interfaces are now a thing, although `Lock.Scope` doesn't ([yet?](https://github.com/dotnet/runtime/issues/111266)) implement `IDisposable` for instance.

So do we treat the duck typing workaround as obsolete and close this?

**Comment by T-Gro** [comment]
Even though the original motivation with code evolution is right, adding `IDisposable` at the source is the correct action and tooling can help with having it stand out.
